### ISSUE TITLE
RevIN

### ISSUE THREAD

**Issue by gdevos010** [lucidrains/iTransformer#6]
I'm not sure if you are looking for additional papers, but RevIN has been found to improve forecasting among many different kinds of models. It's a pretty simple addition where you normalize before the forward pass and then denormalize afterwards.

[Reversible Instance Normalization for Accurate Time-Series Forecasting against Distribution Shift](https://openreview.net/forum?id=cGDAkQo1C0p)
[repo](https://github.com/ts-kim/RevIN/tree/master)

**Comment by lucidrains** [comment]
hey Greg

without reading the paper you linked, is the technique orthogonal to the paper in the repo? i'm only putting in more work if someone tells me the inverted transformer actually works or not

**Comment by lucidrains** [comment]
@gdevos010 do you work with time series data at EpiSci?

**Comment by lucidrains** [comment]
@gdevos010 i see you are working with [medical time series](https://github.com/gdevos010/ml_supervised_learning). tell you what, you run inverted transformer on this, show me that it is at least competitive, and i'll invest the time to read the RevIN paper and figure out how to integrate it, if ideas do not conflict

**Comment by lucidrains** [comment]
we can also move this to discussions, as it is not an issue with the iTrasnformer paper

**Comment by lucidrains** [comment]
@gdevos010 hey Greg, the authors released their code this morning, so likely the paper is going to pan out

i can do a bit of improvisation and get this idea into the repository, just to see if we can push time series sota even further

**Comment by lucidrains** [comment]
ok added

**Comment by lucidrains** [comment]
```python

import torch
from iTransformer import RevIN

x = torch.randn(1, 512, 1024)
rev_in = RevIN(512)
out, reverse_fn = rev_in(x)
y = reverse_fn(out)

assert torch.allclose(x, y)
```

**Comment by lucidrains** [comment]
@gdevos010 do let us know if it works in this context, or not

**Comment by gdevos010** [comment]
Awesome! Thanks for adding. The implementation looks right to me and Im hoping to test out the model this weekend

**Comment by lucidrains** [comment]
@gdevos010 how did it go?

**Comment by gdevos010** [comment]
@lucidrains  I added the model to [darts ](https://github.com/unit8co/darts) framework locally and compared it against some of the other models. It performed worse than other models, but I will keep looking into it will try tuning the hyperparameters.

# input len 250, output len 36
![image](https://github.com/lucidrains/iTransformer/assets/15316026/49254971-b45b-49f6-ae6e-bbf89dcd05c2)
![image](https://github.com/lucidrains/iTransformer/assets/15316026/e34e0ab5-abf6-40e8-b307-2b554f82639f)

# input len 250, output len 100
![image](https://github.com/lucidrains/iTransformer/assets/15316026/6c0c59c2-790d-4e4e-b167-df11a6045acd)
![image](https://github.com/lucidrains/iTransformer/assets/15316026/1bc75986-112c-4d0c-b594-8633e5498a6d)

**Comment by lucidrains** [comment]
@gdevos010 ahh ok, so RevIN has no effect? maybe i should just remove it

**Comment by lucidrains** [comment]
@gdevos010 is this a multivariate dataset?

**Comment by gdevos010** [comment]
I was not comparing RevIN on/off. I only has time for few experiments yesterday and was just testing the model. I will test that today

**Comment by gdevos010** [comment]
Yes, these are multivariate models.

**Comment by lucidrains** [comment]
@gdevos010 nice, thank you for sharing this!

**Comment by lucidrains** [comment]
@gdevos010 want to give the 2d version a try as well? while you are at it?

**Comment by gdevos010** [comment]
@lucidrains I spent some time today trying to get the RevIN working. There's an issue with the tensor shape that I still need to figure out. I believe this has to do with my integration into darts instead of your implementation. 

I also got the 2D version up and running. it is much slower to train. Here is one of the example outputs and I will run it overnight to get more data. 
![image](https://github.com/lucidrains/iTransformer/assets/15316026/87663d6f-1b2e-4de8-8a1f-70ca6c97496f)

**Comment by gdevos010** [comment]
@lucidrains I tested it with 5 more datasets. This hydro energy dataset is hard for a lot of models to learn but itransformer does amazing on it. 
![Backtest_Hydro Energy iTransformer](https://github.com/lucidrains/iTransformer/assets/15316026/29090251-859a-4b5b-b862-ded174ace4ca)
![Backtest_Hydro Energy iTransformer2D](https://github.com/lucidrains/iTransformer/assets/15316026/db2e4128-569d-44de-8e53-74ac47574217)
![Backtest_Hydro Energy NHiTS ()](https://github.com/lucidrains/iTransformer/assets/15316026/a54fc0fe-9840-4715-b5e8-7ce4750fb7c2)

**Comment by lucidrains** [comment]
this is great news! thank you so much Greg!

**Comment by lucidrains** [comment]
for the hydro experiments, was RevIn turned on or off?

**Comment by kashif** [comment]
@gdevos010  so i have a probabilistic variant of iTransform in this PR: https://github.com/awslabs/gluonts/blob/97577be22607ff5c2b5a992c820fd756bf3e2061/examples/iTransformer.ipynb

recall that the RevIN is instance-normalization where the mu and sigma are then used to rescale back the emission on the output side... this technique is related to the original one I believe first done in the DeepAR paper and I have implemented it in the PR where the distribution or its parameters are rescaled back. Also like the Non-stationary transformer I pass the mu and sigma as static covariates (which iTransformer works with unlike dynamic ones) and this technique is also implemented in my PR and I believe is an indirect way (unlike the explicit way in Non-stationary transformer paper) check it out!
